common/src/test/java/io/pravega/common/concurrent/ConcurrentDependentProcessorTests.java

```diff
@@ -0,0 +1,191 @@
+            results.add(proc.add(Collections.singleton(key), () -> {
+                if (!running.compareAndSet(false, true)) {
+                    Assert.fail("Concurrent execution detected.");
+                }
+
+                return thisRun.thenApply(r -> {
+                    running.set(false);
+                    return r;
+                });
+            }));
+
+            if (pr != null) {
+                pr.complete(i - 1);
+            }
+        }
+
+        // Complete the last one.
+        previousRun.get().complete(count - 1);
+
+        for (int i = 0; i < results.size(); i++) {
+            val value = results.get(i).get(TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+            Assert.assertEquals("Unexpected value.", i, (int) value);
+        }
+    }
+
+    /**
+     * Tests the processor using multiple dependency keys for each task. This means that tasks with disjoint dependency
+     * keys may execute in parallel, while the others may not.
+     */
+    @Test
+    public void testMultiDependencyKey() throws Exception {
+        final int key1 = 1;
+        final int key2 = 2;
+        final int key3 = 3;
+        @Cleanup
+        val proc = new ConcurrentDependentProcessor<Integer>(executorService());
+
+        // We setup two individual tasks to begin with.
+        val toRun1 = new CompletableFuture<Integer>();
+        val result1 = proc.add(Collections.singleton(key1), () -> toRun1);
+
+        val toRun2 = new CompletableFuture<Integer>();
+        val result2 = proc.add(Collections.singleton(key2), () -> toRun2);
+        Assert.assertFalse("Not expecting anything to be done yet.", result1.isDone() || result2.isDone());
+
+        // We setup the third task, which depends on both of the original tasks.
+        val result3 = proc.add(Arrays.asList(key1, key2, key3), () -> {
+            Assert.assertTrue("Not expecting third task to execute yet.", result1.isDone() && result2.isDone());
+            return CompletableFuture.completedFuture(3);
+        });
+
+        // Task 4 depends on Key1, which was last used during Task 3.
+        val result4 = proc.add(Collections.singleton(key1), () -> {
+            Assert.assertTrue("Not expecting fourth task to execute yet.", result3.isDone());
+            return CompletableFuture.completedFuture(4);
+        });
+
+        // Task 5 depends on Key3, which was only introduced as a group with Task 3.
+        val result5 = proc.add(Collections.singleton(key3), () -> {
+            Assert.assertTrue("Not expecting fifth task to execute yet.", result3.isDone());
+            return CompletableFuture.completedFuture(5);
+        });
+
+        // Complete the first task. Verify it did complete, but it didn't unblock the third one.
+        toRun1.complete(1);
+        result1.get(TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+        Assert.assertEquals("Unexpected result from first key run.", result1.join(), toRun1.join());
+
+        AssertExtensions.assertThrows(
```

Comment:
There is a variation on this method that just accepts the exception class directly so you don't need an anonymous function